File: runtime/rust/lana-runtime/src/policy.rs
//! Policy evaluation and decision storage, mirroring `runtime/c/policy.c` and
//! `runtime/include/policy.h`.

use std::sync::Arc;

use lana_bytecode::LanaError;
use lana_vm::value::{Value, ValueKind};

use crate::codec;
use crate::sha256;
use crate::store::{Store, store_get, store_put};

#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum PolicyRuleKind {
    ProbabilityAtLeast = 0,
    Equals,
    OrderLessThan,
    Present,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(u32)]
pub enum PolicyOutcome {
    Authorize = 0,
    Refuse,
    RequestMoreEvidence,
}

pub struct PolicyRule {
    pub schema_version: u32,
    pub kind: PolicyRuleKind,
    pub field: Arc<str>,
    pub threshold: f64,
    pub expected: Option<Arc<str>>,
    pub effect: Arc<str>,
}

pub struct Policy {
    pub schema_version: u32,
    pub policy_id: Arc<str>,
    pub rule: PolicyRule,
}

pub struct PolicyEvaluation {
    pub schema_version: u32,
    pub decision_id: u64,
    pub target: Arc<str>,
    pub scope: Arc<str>,
    pub input_revision: u64,
    pub evidence_ids: Option<Arc<str>>,
    pub derivation_ids: Option<Arc<str>>,
    pub relationship_resolution: Option<Arc<str>>,
    pub evaluation_time: u64,
    pub reason: Arc<str>,
    pub requested_evidence: Option<Arc<str>>,
}

pub struct Decision {
    pub schema_version: u32,
    pub decision_id: u64,
    pub policy_id: Arc<str>,
    pub policy_version: [u8; 32],
    pub target: Arc<str>,
    pub scope: Arc<str>,
    pub input_revision: u64,
    pub evidence_ids: Option<Arc<str>>,
    pub derivation_ids: Option<Arc<str>>,
    pub relationship_resolution: Option<Arc<str>>,
    pub outcome: PolicyOutcome,
    pub effect: Option<Arc<str>>,
    pub evaluation_time: u64,
    pub reason: Arc<str>,
    pub requested_evidence: Option<Arc<str>>,
}

fn text_ok(text: Option<&str>) -> bool {
    text.map_or(false, |t| !t.is_empty())
}

fn json_safe(text: Option<&str>) -> bool {
    match text {
        None => true,
        Some(text) => text.bytes().all(|b| b >= 0x20 && b != b'\\' && b != b'"'),
    }
}

fn hex_digest(digest: &[u8; 32]) -> String {
    let mut out = String::with_capacity(64);
    for byte in digest {
        out.push_str(&format!("{:02x}", byte));
    }
    out
}

fn validate_policy(policy: &Policy) -> Result<(), LanaError> {
    if policy.schema_version != 1
        || !text_ok(Some(&policy.policy_id))
        || policy.rule.schema_version != 1
        || !text_ok(Some(&policy.rule.field))
        || !text_ok(Some(&policy.rule.effect))
    {
        return Err(LanaError::Schema);
    }
    if (policy.rule.kind == PolicyRuleKind::ProbabilityAtLeast
        || policy.rule.kind == PolicyRuleKind::OrderLessThan)
        && (policy.rule.threshold < 0.0 || policy.rule.threshold > 1.0)
    {
        return Err(LanaError::InvalidProbability);
    }
    if (policy.rule.kind == PolicyRuleKind::Equals && !text_ok(policy.rule.expected.as_deref()))
        || (policy.rule.kind as u32) > PolicyRuleKind::Present as u32
    {
        return Err(LanaError::Schema);
    }
    Ok(())
}

pub fn policy_version(policy: &Policy) -> Result<[u8; 32], LanaError> {
    validate_policy(policy)?;
    let mut canonical = String::new();
    canonical.push_str(&codec::encode_value(&Value::string(policy.policy_id.clone()))?);
    canonical.push_str(&codec::encode_value(&Value::string(policy.rule.field.clone()))?);
    canonical.push_str(&codec::encode_value(&Value::number(policy.rule.kind as u32 as f64))?);
    canonical.push_str(&codec::encode_value(&Value::number(policy.rule.threshold))?);
    canonical.push_str(&codec::encode_value(&Value::string(
        policy.rule.expected.clone().unwrap_or_else(|| Arc::from("")),
    ))?);
    canonical.push_str(&codec::encode_value(&Value::string(policy.rule.effect.clone()))?);
    Ok(sha256::sha256(canonical.as_bytes()))
}

pub fn policy_store(store: &mut Store, policy: &Policy) -> Result<(), LanaError> {
    validate_policy(policy)?;
    let digest = policy_version(policy)?;
    if !json_safe(Some(&policy.policy_id))
        || !json_safe(Some(&policy.rule.field))
        || !json_safe(policy.rule.expected.as_deref())
        || !json_safe(Some(&policy.rule.effect))
    {
        return Err(LanaError::Schema);
    }
    let version = hex_digest(&digest);
    let key = format!("policy/{}/{}", policy.policy_id, version);
    match store_get(store, &key) {
        Ok(_) => return Err(LanaError::Conflict),
        Err(LanaError::NotFound) => {}
        Err(e) => return Err(e),
    }
    let json = format!(
        "{{\"effect\":\"{}\",\"expected\":\"{}\",\"field\":\"{}\",\"kind\":{},\"policy_id\":\"{}\",\"schema\":1,\"threshold\":{},\"version\":\"{}\"}}",
        policy.rule.effect,
        policy.rule.expected.as_deref().unwrap_or(""),
        policy.rule.field,
        policy.rule.kind as u32,
        policy.policy_id,
        crate::codec::format_17g(policy.rule.threshold),
        version
    );
    store_put(store, &key, &Value::string(Arc::from(json.as_str())))
}

fn condition_matches(rule: &PolicyRule, input: &Value) -> bool {
    let map = match &input.kind {
        ValueKind::Map(map) => map,
        _ => return false,
    };
    let map = map.lock().unwrap();
    let value = match map.get(&rule.field) {
        Some(value) => value,
        None => return false,
    };
    match rule.kind {
        PolicyRuleKind::ProbabilityAtLeast => {
            matches!(value.kind, ValueKind::Number(n) if n >= rule.threshold && n <= 1.0)
        }
        PolicyRuleKind::OrderLessThan => {
            matches!(value.kind, ValueKind::Number(n) if n < rule.threshold && n >= 0.0)
        }
        PolicyRuleKind::Equals => match (&value.kind, &rule.expected) {
            (ValueKind::String(s), Some(expected)) => &**s == &**expected,
            _ => false,
        },
        PolicyRuleKind::Present => true,
    }
}

pub fn policy_evaluate(
    policy: &Policy,
    input: &Value,
    evaluation: &PolicyEvaluation,
) -> Result<Decision, LanaError> {
    if evaluation.schema_version != 1
        || !text_ok(Some(&evaluation.target))
        || !text_ok(Some(&evaluation.scope))
        || !text_ok(Some(&evaluation.reason))
    {
        return Err(LanaError::Schema);
    }
    validate_policy(policy)?;
    let policy_version = policy_version(policy)?;
    let matches = condition_matches(&policy.rule, input);
    Ok(Decision {
        schema_version: 1,
        decision_id: evaluation.decision_id,
        policy_id: policy.policy_id.clone(),
        policy_version,
        target: evaluation.target.clone(),
        scope: evaluation.scope.clone(),
        input_revision: evaluation.input_revision,
        evidence_ids: evaluation.evidence_ids.clone(),
        derivation_ids: evaluation.derivation_ids.clone(),
        relationship_resolution: evaluation.relationship_resolution.clone(),
        outcome: if matches { PolicyOutcome::Authorize } else { PolicyOutcome::RequestMoreEvidence },
        effect: if matches { Some(policy.rule.effect.clone()) } else { None },
        evaluation_time: evaluation.evaluation_time,
        reason: evaluation.reason.clone(),
        requested_evidence: evaluation.requested_evidence.clone(),
    })
}

pub fn policy_store_decision(store: &mut Store, decision: &Decision) -> Result<(), LanaError> {
    if decision.schema_version != 1 {
        return Err(LanaError::Schema);
    }
    let key = format!("decision/{}", decision.decision_id);
    match store_get(store, &key) {
        Ok(_) => return Err(LanaError::Conflict),
        Err(LanaError::NotFound) => {}
        Err(e) => return Err(e),
    }
    if !text_ok(Some(&decision.policy_id))
        || !text_ok(Some(&decision.target))
        || !text_ok(Some(&decision.scope))
        || !text_ok(Some(&decision.reason))
        || !json_safe(Some(&decision.policy_id))
        || !json_safe(Some(&decision.target))
        || !json_safe(Some(&decision.scope))
        || !json_safe(Some(&decision.reason))
        || !json_safe(decision.evidence_ids.as_deref())
        || !json_safe(decision.derivation_ids.as_deref())
        || !json_safe(decision.relationship_resolution.as_deref())
        || !json_safe(decision.effect.as_deref())
        || !json_safe(decision.requested_evidence.as_deref())
    {
        return Err(LanaError::Schema);
    }
    let version = hex_digest(&decision.policy_version);
    let json = format!(
        "{{\"decision_id\":{},\"derivation_ids\":\"{}\",\"effect\":\"{}\",\"evaluation_time\":{},\"evidence_ids\":\"{}\",\"input_revision\":{},\"outcome\":{},\"policy_id\":\"{}\",\"policy_version\":\"{}\",\"relationship_resolution\":\"{}\",\"requested_evidence\":\"{}\",\"reason\":\"{}\",\"scope\":\"{}\",\"target\":\"{}\"}}",
        decision.decision_id,
        decision.derivation_ids.as_deref().unwrap_or(""),
        decision.effect.as_deref().unwrap_or(""),
        decision.evaluation_time,
        decision.evidence_ids.as_deref().unwrap_or(""),
        decision.input_revision,
        decision.outcome as u32,
        decision.policy_id,
        version,
        decision.relationship_resolution.as_deref().unwrap_or(""),
        decision.requested_evidence.as_deref().unwrap_or(""),
        decision.reason,
        decision.scope,
        decision.target
    );
    store_put(store, &key, &Value::string(Arc::from(json.as_str())))
}

pub fn policy_replay(
    policy: &Policy,
    input: &Value,
    evaluation: &PolicyEvaluation,
    expected: &Decision,
) -> Result<Decision, LanaError> {
    if expected.schema_version != 1 {
        return Err(LanaError::Schema);
    }
    let replayed = policy_evaluate(policy, input, evaluation)?;
    if replayed.policy_version != expected.policy_version
        || replayed.outcome != expected.outcome
        || replayed.decision_id != expected.decision_id
        || replayed.input_revision != expected.input_revision
    {
        return Err(LanaError::Integrity);
    }
    Ok(replayed)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_policy() -> Policy {
        Policy {
            schema_version: 1,
            policy_id: Arc::from("p1"),
            rule: PolicyRule {
                schema_version: 1,
                kind: PolicyRuleKind::ProbabilityAtLeast,
                field: Arc::from("p"),
                threshold: 0.5,
                expected: None,
                effect: Arc::from("grant"),
            },
        }
    }

    #[test]
    fn version_is_stable() {
        let policy = sample_policy();
        let a = policy_version(&policy).unwrap();
        let b = policy_version(&policy).unwrap();
        assert_eq!(a, b);
    }

    #[test]
    fn evaluate_authorizes_when_above_threshold() {
        let policy = sample_policy();
        let mut map = lana_vm::value::Map::new(1);
        map.set(Arc::from("p"), Value::number(0.75), false).unwrap();
        let input = Value::map(Arc::new(std::sync::Mutex::new(map)));
        let evaluation = PolicyEvaluation {
            schema_version: 1,
            decision_id: 1,
            target: Arc::from("t"),
            scope: Arc::from("s"),
            input_revision: 0,
            evidence_ids: None,
            derivation_ids: None,
            relationship_resolution: None,
            evaluation_time: 0,
            reason: Arc::from("r"),
            requested_evidence: None,
        };
        let decision = policy_evaluate(&policy, &input, &evaluation).unwrap();
        assert_eq!(decision.outcome, PolicyOutcome::Authorize);
        assert_eq!(decision.effect.as_deref(), Some("grant"));
    }
}

File: runtime/c/policy.c
#include "policy.h"

#include "codec.h"
#include "data.h"
#include "sha256.h"
#include "store.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static bool text_ok(const char *text) { return text != NULL && text[0] != '\0'; }

static void hex_digest(const unsigned char digest[32], char output[65]) {
    static const char hex[] = "0123456789abcdef"; size_t index;
    for (index = 0u; index < 32u; ++index) { output[index * 2u] = hex[digest[index] >> 4u]; output[index * 2u + 1u] = hex[digest[index] & 15u]; }
    output[64] = '\0';
}

static bool json_safe(const char *text) {
    size_t index;
    if (text == NULL) return true;
    for (index = 0u; text[index] != '\0'; ++index) if ((unsigned char)text[index] < 0x20u || text[index] == '\\' || text[index] == '"') return false;
    return true;
}

static LanaError validate_policy(const LanaPolicy *policy) {
    if (policy == NULL || policy->struct_size < sizeof(*policy) || policy->schema_version != 1u ||
        !text_ok(policy->policy_id) || policy->rule.struct_size < sizeof(policy->rule) ||
        policy->rule.schema_version != 1u || !text_ok(policy->rule.field) || !text_ok(policy->rule.effect))
        return LANA_ERR_SCHEMA;
    if ((policy->rule.kind == LANA_POLICY_PROBABILITY_AT_LEAST ||
         policy->rule.kind == LANA_POLICY_ORDER_LESS_THAN) &&
        (policy->rule.threshold < 0.0 || policy->rule.threshold > 1.0)) return LANA_ERR_INVALID_PROBABILITY;
    if ((policy->rule.kind == LANA_POLICY_EQUALS && !text_ok(policy->rule.expected)) ||
        policy->rule.kind > LANA_POLICY_PRESENT) return LANA_ERR_SCHEMA;
    return LANA_OK;
}

LanaError lana_policy_version(const LanaPolicy *policy, unsigned char out_version[32]) {
    LanaBuffer canonical = {0}; LanaError error;
    if (out_version == NULL) return LANA_ERR_INVALID_STATE;
    error = validate_policy(policy); if (error != LANA_OK) return error;
    error = lana_codec_encode_value(&canonical, lana_value_string((char *)policy->policy_id));
    if (error == LANA_OK) error = lana_codec_encode_value(&canonical, lana_value_string((char *)policy->rule.field));
    if (error == LANA_OK) error = lana_codec_encode_value(&canonical, lana_value_number((double)policy->rule.kind));
    if (error == LANA_OK) error = lana_codec_encode_value(&canonical, lana_value_number(policy->rule.threshold));
    if (error == LANA_OK) error = lana_codec_encode_value(&canonical, lana_value_string((char *)(policy->rule.expected == NULL ? "" : policy->rule.expected)));
    if (error == LANA_OK) error = lana_codec_encode_value(&canonical, lana_value_string((char *)policy->rule.effect));
    if (error == LANA_OK) lana_sha256(canonical.data, canonical.length, out_version);
    free(canonical.data); return error;
}

LanaError lana_policy_store(LanaStore *store, LanaVM *vm, const LanaPolicy *policy) {
    unsigned char digest[32]; char version[65], key[160], json[1024]; Value existing, value; LanaError error;
    if (store == NULL || vm == NULL) return LANA_ERR_INVALID_STATE;
    error = validate_policy(policy); if (error != LANA_OK) return error;
    if ((error = lana_policy_version(policy, digest)) != LANA_OK) return error;
    if (!json_safe(policy->policy_id) || !json_safe(policy->rule.field) || !json_safe(policy->rule.expected) || !json_safe(policy->rule.effect)) return LANA_ERR_SCHEMA;
    hex_digest(digest, version);
    (void)snprintf(key, sizeof(key), "policy/%s/%s", policy->policy_id, version);
    error = lana_store_get(store, vm, key, &existing);
    if (error == LANA_OK) return LANA_ERR_CONFLICT;
    if (error != LANA_ERR_NOT_FOUND) return error;
    (void)snprintf(json, sizeof(json), "{\"effect\":\"%s\",\"expected\":\"%s\",\"field\":\"%s\",\"kind\":%u,\"policy_id\":\"%s\",\"schema\":1,\"threshold\":%.17g,\"version\":\"%s\"}",
                   policy->rule.effect, policy->rule.expected == NULL ? "" : policy->rule.expected,
                   policy->rule.field, (unsigned)policy->rule.kind, policy->policy_id, policy->rule.threshold, version);
    if (strlen(json) >= sizeof(json) - 1u) return LANA_ERR_LIMIT;
    value = lana_value_string(json); return lana_store_put(store, key, value);
}

static bool condition_matches(const LanaPolicyRule *rule, const Value *input) {
    Value value;
    if (input == NULL || input->type != VAL_MAP || lana_map_get(input->as.map, rule->field, &value) != LANA_OK) return false;
    switch (rule->kind) {
        case LANA_POLICY_PROBABILITY_AT_LEAST: return value.type == VAL_NUMBER && value.as.number >= rule->threshold && value.as.number <= 1.0;
        case LANA_POLICY_ORDER_LESS_THAN: return value.type == VAL_NUMBER && value.as.number < rule->threshold && value.as.number >= 0.0;
        case LANA_POLICY_EQUALS: return value.type == VAL_STRING && strcmp(value.as.string, rule->expected) == 0;
        case LANA_POLICY_PRESENT: return true;
        default: return false;
    }
}

LanaError lana_policy_evaluate(const LanaPolicy *policy, const Value *input,
                               const LanaPolicyEvaluation *evaluation, LanaDecision *out) {
    LanaError error; bool matches;
    if (out == NULL || evaluation == NULL || evaluation->struct_size < sizeof(*evaluation) || evaluation->schema_version != 1u ||
        !text_ok(evaluation->target) || !text_ok(evaluation->scope) || !text_ok(evaluation->reason)) return LANA_ERR_SCHEMA;
    error = validate_policy(policy); if (error != LANA_OK) return error;
    memset(out, 0, sizeof(*out)); out->struct_size = sizeof(*out); out->schema_version = 1u;
    out->decision_id = evaluation->decision_id; out->policy_id = policy->policy_id;
    out->target = evaluation->target; out->scope = evaluation->scope; out->input_revision = evaluation->input_revision;
    out->evidence_ids = evaluation->evidence_ids; out->derivation_ids = evaluation->derivation_ids;
    out->relationship_resolution = evaluation->relationship_resolution; out->evaluation_time = evaluation->evaluation_time;
    out->reason = evaluation->reason; out->requested_evidence = evaluation->requested_evidence;
    error = lana_policy_version(policy, out->policy_version); if (error != LANA_OK) return error;
    matches = condition_matches(&policy->rule, input);
    out->outcome = matches ? POLICY_OUTCOME_AUTHORIZE : POLICY_OUTCOME_REQUEST_MORE_EVIDENCE;
    out->effect = matches ? policy->rule.effect : NULL;
    return LANA_OK;
}

LanaError lana_policy_store_decision(LanaStore *store, LanaVM *vm, const LanaDecision *decision) {
    char key[64], json[4096], version[65]; Value existing, value; LanaError error;
    if (store == NULL || vm == NULL || decision == NULL || decision->struct_size < sizeof(*decision) || decision->schema_version != 1u) return LANA_ERR_SCHEMA;
    (void)snprintf(key, sizeof(key), "decision/%llu", (unsigned long long)decision->decision_id);
    error = lana_store_get(store, vm, key, &existing);
    if (error == LANA_OK) return LANA_ERR_CONFLICT;
    if (error != LANA_ERR_NOT_FOUND) return error;
    if (!text_ok(decision->policy_id) || !text_ok(decision->target) || !text_ok(decision->scope) ||
        !text_ok(decision->reason) || !json_safe(decision->policy_id) || !json_safe(decision->target) ||
        !json_safe(decision->scope) || !json_safe(decision->reason) || !json_safe(decision->evidence_ids) ||
        !json_safe(decision->derivation_ids) || !json_safe(decision->relationship_resolution) ||
        !json_safe(decision->effect) || !json_safe(decision->requested_evidence)) return LANA_ERR_SCHEMA;
    hex_digest(decision->policy_version, version);
    (void)snprintf(json, sizeof(json),
                   "{\"decision_id\":%llu,\"derivation_ids\":\"%s\",\"effect\":\"%s\",\"evaluation_time\":%llu,\"evidence_ids\":\"%s\",\"input_revision\":%llu,\"outcome\":%u,\"policy_id\":\"%s\",\"policy_version\":\"%s\",\"relationship_resolution\":\"%s\",\"requested_evidence\":\"%s\",\"reason\":\"%s\",\"scope\":\"%s\",\"target\":\"%s\"}",
                   (unsigned long long)decision->decision_id, decision->derivation_ids == NULL ? "" : decision->derivation_ids,
                   decision->effect == NULL ? "" : decision->effect, (unsigned long long)decision->evaluation_time,
                   decision->evidence_ids == NULL ? "" : decision->evidence_ids, (unsigned long long)decision->input_revision,
                   (unsigned)decision->outcome, decision->policy_id, version,
                   decision->relationship_resolution == NULL ? "" : decision->relationship_resolution,
                   decision->requested_evidence == NULL ? "" : decision->requested_evidence,
                   decision->reason, decision->scope, decision->target);
    if (strlen(json) >= sizeof(json) - 1u) return LANA_ERR_LIMIT;
    value = lana_value_string(json); return lana_store_put(store, key, value);
}

LanaError lana_policy_replay(const LanaPolicy *policy, const Value *input,
                             const LanaPolicyEvaluation *evaluation,
                             const LanaDecision *expected, LanaDecision *out) {
    LanaDecision replayed; LanaError error;
    if (expected == NULL || expected->struct_size < sizeof(*expected) || expected->schema_version != 1u) return LANA_ERR_SCHEMA;
    if ((error = lana_policy_evaluate(policy, input, evaluation, &replayed)) != LANA_OK) return error;
    if (memcmp(replayed.policy_version, expected->policy_version, sizeof(replayed.policy_version)) != 0 ||
        replayed.outcome != expected->outcome || replayed.decision_id != expected->decision_id ||
        replayed.input_revision != expected->input_revision) return LANA_ERR_INTEGRITY;
    if (out != NULL) *out = replayed;
    return LANA_OK;
}

void lana_decision_free(LanaDecision *decision) { (void)decision; }
